src/Build/BackEnd/Components/ProjectCache/ProjectCacheService.cs

```diff
@@ -218,18 +229,45 @@ private IEnumerable<ProjectCacheDescriptor> GetProjectCacheDescriptors(ProjectIn
                     ProjectCacheException.ThrowForErrorLoggedInsideTheProjectCache("ProjectCacheInitializationFailed");
                 }
 
-                return new ProjectCachePlugin(pluginTypeName, pluginInstance);
+                FileAccessManager.HandlerRegistration handlerRegistration = _fileAccessManager.RegisterHandlers(
+                    (buildRequest, fileAccessData) =>
+                    {
+                        // TODO: Filter out projects which do not configure this plugin
+                        FileAccessContext fileAccessContext = GetFileAccessContext(buildRequest);
+                        pluginInstance.HandleFileAccess(fileAccessContext, fileAccessData);
+                    },
+                    (buildRequest, processData) =>
+                    {
+                        // TODO: Filter out projects which do not configure this plugin
+                        FileAccessContext fileAccessContext = GetFileAccessContext(buildRequest);
+                        pluginInstance.HandleProcess(fileAccessContext, processData);
+                    });
+
+                return new ProjectCachePlugin(pluginTypeName, pluginInstance, handlerRegistration);
             }
             catch (Exception e)
             {
-                return new ProjectCachePlugin(pluginTypeName, Instance: null, ExceptionDispatchInfo.Capture(e));
+                return new ProjectCachePlugin(pluginTypeName, Instance: null, HandlerRegistration: null, ExceptionDispatchInfo.Capture(e));
             }
             finally
             {
                 MSBuildEventSource.Log.ProjectCacheBeginBuildStop(pluginTypeName);
             }
         }
 
+        private FileAccessContext GetFileAccessContext(BuildRequest buildRequest)
+        {
+            BuildRequestConfiguration configuration = _configCache[buildRequest.ConfigurationId];
+
+            var globalProperties = new Dictionary<string, string>(configuration.GlobalProperties.Count, StringComparer.OrdinalIgnoreCase);
```

Comment:
I'd love to avoid the allocations here. Personally I'd like to just expose `configuration.GlobalProperties` directly, but that's not a public type.

It would be possible to create a cache of dictionaries if we think that the unique set of global properties is small compared to the number of projects, and if the dictionary operations are fast enough (comparer impl). I'm not convinced it worth it, but I'm also not convinced it's not.

Reply:
You mean that PropertyDictionary isn't public? It's an IDictionary; can you just use that?

The main reasons I know of for global properties to change is for multitargeting and with the MSBuild task. The former should be a pretty small space of options, but I'll admit the latter...not so much.

Reply:
`PropertyDictionary` is public, but is an `IDictionary<string, ProjectPropertyInstance>` and `ProjectPropertyInstance` is mutable.